README.md
# SSAT
Shell Script Automated Tester (unit testing executable files)


## Files

ssat.sh: the main executable (bash script)

ssat-api.sh: API function definitions that user script can use.

util/: utility binaries for extended features usually implemented for nnstreamer/nntrainer community.


debian/changelog
ssat (1.4.0) stable; urgency=medium

  * Vangrind support added.

 -- MyungJoo Ham <myungjoo.ham@samsung.com>  Thu, 17 Aug 2023 17:21:00 +0900

ssat (1.2.0) stable; urgency=medium

  * Progress and error logs revised.

 -- MyungJoo Ham <myungjoo.ham@samsung.com>  Mon, 22 Feb 2021 17:23:00 +0900

ssat (1.1.0) stable; urgency=medium

  * A lot of updates have been done (new features)

 -- MyungJoo Ham <myungjoo.ham@samsung.com>  Mon, 16 Mar 2020 19:39:00 +0900

ssat (1.0.0) stable; urgency=medium

  * Initial debian dist files added

 -- MyungJoo Ham <myungjoo.ham@samsung.com>  Tue, 30 Oct 2018 17:10:00 +0900


ssat.sh
#!/usr/bin/env bash
##
## @file ssat.sh
## @author MyungJoo Ham <myungjoo.ham@gmail.com>
## @date Jun 22 2018
## @brief This executes test groups and reports aggregated test results.
## @return 0 if all PASSED. Positive if some FAILED.
## @todo Separate GStreamer related functions as plugins
##
## This uses sed, date, cmp
##
## If there is no arguments specified, this will search for all "runTest.sh" in
## the subdirectory of this file and regard them as the test groups.
##
## If a testgroup (runTest.sh) returns 0 while there are failed testcase,
## it implies that the failed testcases may be ignored and it's good to go.
##
## If --help or -h is given, this will show detailed description.

##
## @mainpage SSAT
## @section intro        Introduction
## - Introduction     :  Shell Script Automated Tester
## @section Program      Program Name
## - Program Name     :  ssat
## - Program Details  :  SSAT is a software testing framework for test cases written in BASH shell scripts.
##   It can search for test scripts recursively from a given path and summarize the test results.
##   If there is any "critical" fail, ssat will return non-zero values on its exit.
## @section INOUTPUT     Input/output data
## - INPUT            :  Test Cases (If not supplied, the current path is the root of test cases)
## - OUTPUT           :  Summary of test results to stdout. Exit code of 0 if success, non-zero if not success.
## @section CREATEINFO   Code information
## - Initial date     :  2018/06/22
## - Version          :  1.4.0

TARGET=$(pwd)
TARGET_ASSIGNED=0
BASEPATH=`dirname "$0"`
BASENAME=`basename "$0"`
TESTCASE="runTest.sh"
SUMMARYFILENAME=""
VERSION="1.4.0"

#
SILENT=1
PROGRESSLOGLEVEL=0
COUNTNEGATIVE=0
COUNTNEGATIVEPOSTFIX=""
VALGRIND=0
VALGRIND_SUPPRESSION=""
date=`date +"%b %d %Y"`

## @fn createTemplate()
## @brief Generate runTest template file
##
## Note that the generated template has no license.
## The SSAT user may put their own license for the generated files.
## I hereby grant the right to relicense the generated files.
function createTemplate() {
	if [[ -f "runTest.sh" ]]
	then
		printf "Cannot create runTest.sh here. The file already exists at $(pwd).\n\n"
		exit 1
	fi

	echo -e "#!/usr/bin/env bash\n\
##\n\
## @file runTest.sh\n\
## @author MyungJoo Ham <myungjoo.ham@gmail.com>\n\
## @date ${date}\n\
## @brief This is a template file for SSAT test cases. You may designate your own license.\n\
#\n\
if [[ \"\$SSATAPILOADED\" != \"1\" ]]\n\
then\n\
	SILENT=0\n\
	INDEPENDENT=1\n\
	search=\"ssat-api.sh\"\n\
	source \$search\n\
	printf \"\${Blue}Independent Mode\${NC}\\n\"\n\
fi\n\
testInit \$1 # You may replace this with Test Group Name\n\
\n\
#testResult 1 T1 \"Dummy Test1\"\n\
#callTestSuccess gst-launch-1.0 \"-q videotestsrc ! videoconvert ! autovideosink\" T2 \"This may run indefinitely\"\n\
#callCompareTest golden.log executeResult.log T3 \"The two files must be same\" 0\n\
\n\
report\n" > runTest.sh
chmod a+x runTest.sh

	exit 0
}

# Handle arguments
POSITIONAL=()
while [[ $# -gt 0 ]]
do
	key="$1"
	case $key in
	-h|--help)
		printf "usage: ${BASENAME} [--help] [<path>] [--testcase <filename>] [--nocolor] [--showstdout] [--createtemplate] [--countnegative <postfix>] [--enable-valgrind] [--valgrind-suppression <filepath>] [--version]\n\n"
		printf "These are common ${Red}ssat${NC} commands used:\n\n"
		printf "Test all test-groups in the current ($(pwd)) directory, recursively\n"
		printf "    (no options specified)\n"
		printf "    $ ${BASENAME}\n"
		printf "\n"
		printf "Test all test-groups in the specified directory, recursively\n"
		printf "    <path>\n"
		printf "    $ ${BASENAME} /home/username/test\n"
		printf "    If there are multiple paths, the last one will be used\n"
		printf "\n"
		printf "Search for \"filename\" as the testcase scripts\n"
		printf "    --testcase or -t\n"
		printf "    $ ${BASENAME} --testcase cases.sh\n"
		printf "    Search for cases.sh instead of runTest.sh\n"
		printf "\n"
		printf "Do not emit colored text\n"
		printf "    --nocolor or -n\n"
		printf "\n"
		printf "Show stdout of test cases\n"
		printf "    --showstdout or -s\n"
		printf "\n"
		printf "Create a template 'runTest.sh' test group at your current directory\n"
		printf "    --createtemplate or -c\n"
		printf "\n"
		printf "Show progress during execution\n"
		printf "    --progress or -p or -p=(0,1,9)\n"
		printf "        '0' : Do not print logs in progress. If -p is not given, -p=0 is assumed.\n"
		printf "        '1' : Print test group names only in progress.\n"
		printf "        '2-9' : Print all logs in progress. If -p is given without numbers, -p=9 is used.\n"
		printf "     $ ${BASENAME} -p=1\n"
		printf "     $ ${BASENAME} --progress=9 (equal to --progress) \n"
		printf "\n"
		printf "Enable valgrind to perform memcheck\n"
		printf "    --enable-valgrind or -vg\n"
		printf "\n"
		printf "Suppress valgrind errors with the given suppression file\n"
		printf "    --valgrind-suppression <path to the suppression file>\n"
		printf "    or \n"
		printf "    -vs <path to the suppression file>\n"
		printf "\n"
		printf "Shows this message\n"
		printf "    --help or -h\n"
		printf "    $ ${BASENAME} --help \n"
		printf "\n"
		printf "Count negative test cases with the given postfix\n"
		printf "    --countnegative or -cn\n"
		printf "    $ ${BASENAME} --countnegative _n\n"
		printf "    $ ${BASENAME} -cn _n\n"
		printf "\n"
		printf "Write result summary as a file\n"
		printf "    --summary <filename>\n"
		printf "\n"
		printf "Show the version\n"
		printf "    --version or -v\n"
		printf "\n\n"
		exit 0
	;;
	-n|--nocolor)
	nocolor=1
	shift
	;;
	-t|--testcase)
	TESTCASE="$2"
	shift
	shift
	;;
	-cn|--countnegative)
	COUNTNEGATIVE=1
	COUNTNEGATIVEPOSTFIX="$2"
	if [[ "${COUNTNEGATIVEPOSTFIX}" == "" ]]
	then
		printf "${BASENAME} -cn or --countnegative requires postfix.\n\n"
		exit -2
	fi
	shift
	shift
	;;
	-s|--showstdout)
	SILENT=0
	shift
	;;
	-c|--createtemplate)
	createTemplate
	shift
	;;
	-p|-p=*|--progress|--progress=*)
	if [[ $key == "-p" || $key == "--progress" ]]
	then
	    PROGRESSLOGLEVEL=9
	    printf "Progress Log level is not given. Print all logs in progress.\n"
	else
	    PROGRESSLOGLEVEL=${key#*=}
	    printf "Given progress log level is ${PROGRESSLOGLEVEL}.\n"
	fi
	shift
	;;
	-vg|--enable-valgrind)
	VALGRIND=1
	shift
	;;
	-vs|--valgrind-suppression)
	VALGRIND_SUPPRESSION=" --suppressions=$2 "
	shift
	shift
	;;
	--summary)
	SUMMARYFILENAME="$2"
	shift
	shift
	;;
	--version|-v)
	printf "${VERSION}\n"
	exit 0
	shift
	;;
	*) # Unknown, which is probably target (the path to root-dir of test groups).
	# If this is the second occurrence, ignore it.
	# Assume that the previous string is path and the later string is an invalid argument.
	if [ $TARGET_ASSIGNED -eq 0 ]
	then
		TARGET="$1"
		TARGET_ASSIGNED=1
	fi
	shift
	esac
done

source ${BASEPATH}/ssat-api.sh

if [[ "${#TARGET}" -eq "0" ]]
then
	TARGET="."
fi

TNtc=0
TNtcpass=0
TNtcfail=0
TNtcignore=0
TNtcneg=0
TNgroup=0
TNgrouppass=0
TNgroupfail=0
log=""
groupLog=""

while read -d $'\0' file
do
	CASEBASEPATH=`dirname "$file"`
	CASENAME=`basename "$CASEBASEPATH"`
	Ntc=0
	Npass=0
	Nfail=0
	Nneg=0
	tmpfile=$(mktemp)

	if [[ ${PROGRESSLOGLEVEL} -ge 1 ]]; then
		printf "[Starting] $CASENAME\n"
	fi
	pushd $CASEBASEPATH > /dev/null
	output=$(. $file $CASEBASEPATH)
	retcode=$?
	popd > /dev/null

	logfile="${output##*$'\n'}"
	resultlog=$(<$logfile)
	effectiveOutput=`printf "$resultlog" | sed '$d'`
	log="$log$effectiveOutput\n"

	lastline=`printf "${resultlog}" | sed '$!d'`
	IFS=,
	set $lastline
	Ntc=$1
	Npass=$2
	Nfail=$3
	Nignore=$4
	Nneg=$5
	unset IFS

	TNtc=$((TNtc+Ntc))
	TNtcpass=$((TNtcpass+Npass))
	TNtcfail=$((TNtcfail+Nfail))
	TNtcignore=$((TNtcignore+Nignore))
	TNtcneg=$((TNtcneg+Nneg))

	TNgroup=$((TNgroup+1))
	if [[ "$retcode" -eq "0" ]]
	then
		TNgrouppass=$((TNgrouppass+1))
		groupLog="${groupLog}${LightGreen}[PASSED]${NC} ${Blue}${CASENAME}${NC} ($Npass passed among $Ntc cases)\n"
	else
		TNgroupfail=$((TNgroupfail+1))
		groupLog="${groupLog}${Red}[FAILED]${NC} ${Blue}${CASENAME}${NC} ($Npass passed among $Ntc cases)\n"
	fi

	printf "$log\n"
	log=""
done < <(find $TARGET -name $TESTCASE -print0)

printf "\n\n==================================================\n\n"

printf "==================================================\n\n"
printf "$groupLog"
printf "==================================================\n"

ADDITIONALSTRING=""
ADDITIONALSUMMARY=""
if (( ${COUNTNEGATIVE} == 1 ))
then
	total=$((TNtcpass+TNtcfail+TNtcignore))
	pos=$((total-TNtcneg))
	ADDITIONALSTRING="${ADDITIONALSTRING} | Positive: ${pos} / Negative: ${TNtcneg}"
	ADDITIONALSUMMARY="${ADDITIONALSUMMARY}, negative=${TNtcneg}"
fi

if [ "${SUMMARYFILENAME}" != "" ]
then
	echo "passed=${TNtcpass}, failed=${TNtcfail}, ignored=${TNtcignore}${ADDITIONALSUMMARY}" > "${SUMMARYFILENAME}"
fi
if (( ${TNgroupfail} == 0 ))
then
	printf "${LightGreen}[PASSED] ${Blue}All Test Groups (${TNgroup}) Passed!${NC}\n"
	printf "         TC Passed: ${TNtcpass} / Failed: ${TNtcfail} / Ignored: ${TNtcignore} ${ADDITIONALSTRING}\n\n";
	exit 0
else
	printf "${Red}[FAILED] ${Purple}There are failed test groups! (${TNgroupfail})${NC}\n"
	printf "         TC Passed: ${TNtcpass} / Failed: ${TNtcfail} / Ignored: ${TNtcignore} ${ADDITIONALSTRING}\n\n";
	exit 1
fi
# gather reports & publish them.


exampleCase/gstTestBackground/runTest.sh
#!/usr/bin/env bash

if [[ "$SSATAPILOADED" != "1" ]]
then
	SILENT=0
	INDEPENDENT=1
	search="ssat-api.sh"
	source $search

	retcode=$?
	count=0
	while (( ${retcode} != 0 ))
	do
		count=$((count+1))
		if (( ${count} > 5 ))
		then
			echo "Cannot find ssat-api.sh"
			exit 1
		fi

		search="../${search}"
		source $search
		retcode=$?
	done
	printf "${Blue}Independent Mode${NC}\n"
fi

testInit $1

command -v gst-launch-1.0 || report
command -v gst-inspect-1.0 || report

filename=$(mktemp)
gstTestBackground "videotestsrc num-buffers=5 is-live=true  ! video/x-raw,width=64,height=48,framerate=1/1 ! filesink location=${filename}" GBKG1 0 0 5
testResult 1 GBKG1-AFT "After launching the background" 0
wait $pid
testResult 1 GBKG1-CMP "After the background is completed" 0

gst-inspect-1.0 udpsrc &> /dev/null
val=$?
if [ "${val}" == "0" ]; then
	gstTestBackground "udpsrc ! video/x-raw,width=64,height=48,framerate=1/1 ! fakesink async=false" GBKG2 0 0 5
	testResult 1 GBKG2-AFT "After launching the background" 0
	kill $pid
	testResult 1 GBKG2-CMP "After the background is killed" 0

	gstTestBackground "udpsrc ! video/x-raw,width=64,height=48,framerate=1/1 ! fakesink" GBKG3 0 1 5
	testResult 1 GBKG3-AFT "After launching the background" 0
	kill $pid
	testResult 1 GBKG3-CMP "After the background is killed" 0
else
	testResult 0 GBKG2 "udpsrc-based gst background test failed: udpsrc not found: gst-inspect returns ${val}." 1
fi


report


ssat-api.sh
#!/usr/bin/env bash
##
## @file ssat-api.sh
## @author MyungJoo Ham <myungjoo.ham@gmail.com>
## @date Jun 22 2018
## @brief This is API set for SSAT (Shell Script Automated Tester)
##

if [[ "$nocolor" != "1" ]]
then
	Black='\033[0;30m'
	DarkGray='\033[1;30m'
	Red='\033[0;31m'
	LightRed='\033[1;31m'
	Green='\033[0;32m'
	LightGreen='\033[1;32m'
	Orange='\033[0;33m'
	Yellow='\033[1;33m'
	Blue='\033[0;34m'
	LightBlue='\033[1;34m'
	Purple='\033[0;35m'
	LightPurple='\033[1;35m'
	Cyan='\033[0;36m'
	LightCyan='\033[1;36m'
	LightGray='\033[0;37m'
	White='\033[1;37m'
	NC='\033[0m'
else
	Black=''
	DarkGray=''
	Red=''
	LightRed=''
	Green=''
	LightGreen=''
	Orange=''
	Yellow=''
	Blue=''
	LightBlue=''
	Purple=''
	LightPurple=''
	Cyan=''
	LightCyan=''
	LightGray=''
	White=''
	NC=''
fi


ResultLog=""

# Platform dependent variables
KernelName=$(uname -s)
if [[ "${KernelName}" == "Darwin" ]]; then
	StatCmd_GetSize="stat -f %z"
	SO_EXT="dylib"
else
	StatCmd_GetSize="stat --printf=%s"
	SO_EXT="so"
fi

## @fn writef()
## @private
## @param $1 the string to be printed.
## @brief Prepare report result
function writef() {
	if [[ "${SILENT}" == "0" ]]
	then
		printf "$1\n"
	fi
	ResultLog="$ResultLog$1\n"
}

## @fn report()
## @brief Report results of a test group (a "runTest.sh" in a testee directory)
function report() {
	if (( ${_fail} == 0 && ${_criticalFail} == 0 ))
	then
		writef "${Green}==================================================${NC}"
		writef "${LightGreen}[PASSED]${NC} Test Group $_group ${Green}Passed${NC}"
	else
		if (( ${_criticalFail} > 0 ))
		then
			writef "${Green}==================================================${NC}"
			writef "${Red}[FAILED]${NC} Test Group $_group has ${Red}failed cases ($_fail)${NC}"
		else
			writef "${Green}==================================================${NC}"
			writef "${LightGreen}[PASSED]${NC} Test Group $_group has ${Red}failed cases ($_fail), but they are ignorable cases and not critical.${NC}"
		fi
	fi

	if [[ "$INDEPENDENT" -eq "1" ]]
	then
	# do nothing
		echo ""
	else
		_ignore=$((_fail-_criticalFail))
		_fail=${_criticalFail}
		if [[ "${COUNTNEGATIVE}" -eq "1" ]]
		then
			writef "${_cases},${_pass},${_fail},${_ignore},${_neg}"
		else
			writef "${_cases},${_pass},${_fail},${_ignore}"
		fi
		echo "${ResultLog}" > ${_filename}
		printf "\n${_filename}\n"
	fi

	if (( ${_criticalFail} > 0 ))
	then
		exit 1
	else
		exit 0
	fi
}

## @fn testInit()
## @brief Initialize runTest.sh shell test case
function testInit() {
	_pass=0
	_fail=0
	_criticalFail=0
	_cases=0
	_neg=0
	_filename=$(mktemp)
	_group=`basename "$1"`
	if [[ "${#_group}" -eq "0" ]]
	then
		_group="(Unspecified)"
	fi

	writef "${Green}==================================================${NC}"
	writef "    Test Group ${Green}$_group${NC} Starts."
}

## @fn testResult()
## @brief Write Test Log
## @param $1 1 = success / 0 = fail ($5 is not 1)
## @param $2 test case ID (short string)
## @param $3 test case description
## @param $4 set 1 if this is not critical (don't care if it's pass or fail)_
## @param $5 set 1 if $1==0 is success and $1!=0 is fail.
function testResult() {
	if [[ ${PROGRESSLOGLEVEL} -gt 1 ]]
	then
		echo "Case ${2}(${3}) report ${1}" >&2
	fi

	_cases=$((_cases+1))
	_good=0
	if [[ "${5}" -eq "1" ]]; then
		if [[ "${1}" -eq "0" ]]; then
			_good=1
		fi
	else
		if [[ "${1}" -eq "1" ]]; then
			_good=1
		fi
	fi

	if [[ "${COUNTNEGATIVE}" -eq "1" ]]
	then
		if [[ "${2}\n" =~ "${COUNTNEGATIVEPOSTFIX}\n" ]]
		then
			_neg=$((_neg+1))
		fi
	fi

	if [[ "${_good}" -eq "1" ]]
	then
		writef "${LightGreen}[PASSED]${NC} ${Green}$2${NC}:$3${NC}"
		_pass=$((_pass+1))
	else
		_fail=$((_fail+1))
		if [[ "${4}" == "1" ]]
		then
			writef "${Purple}[IGNORED] $2${NC}:${Purple}$3${NC}"
		else
			writef "${Red}[FAILED][Critical] $2${NC}:${Purple}$3${NC}"
			_criticalFail=$((_criticalFail+1))
		fi
	fi
}

## @fn callTestSuccess()
## @brief Call Test Case (a shell script), expected exit = 0
## @param $1 Full path to the executable (e.g., ~/script/a1.sh)
## @param $2 Full string of the arguments to $1 (e.g., "-q -n --dryrun")
## @param $3 test case ID
## @param $4 test case description
## @param $5 set 1 if this is not critical (don't care if it's pass or fail)_
function callTestSuccess() {
	callutput=$(. $1 $2)
	retcode=$?
	if (( ${retcode} == 0 ))
	then
		testResult 1 "$3" "$4" $5
	else
		testResult 0 "$3" "$4 ret($retcode)" $5
	fi
}

## @fn callTestFail()
## @brief Call Test Case (a shell script), expected exit != 0
## @param $1 Full path to the executable (e.g., ~/script/a1.sh)
## @param $2 Full string of the arguments to $1 (e.g., "-q -n --dryrun")
## @param $3 test case ID
## @param $4 test case description
## @param $5 set 1 if this is not critical (don't care if it's pass or fail)_
function callTestFail() {
	callutput=$(. $1 $2)
	retcode=$?
	if (( ${retcode} != 0 ))
	then
		testResult 1 "$3" "$4 ret($retcode)" $5
	else
		testResult 0 "$3" "$4" $5
	fi
}

## @fn callTestExitEq()
## @brief Call Test Case (a shell script), expected exit == $5
## @param $1 Full path to the executable (e.g., ~/script/a1.sh)
## @param $2 Full string of the arguments to $1 (e.g., "-q -n --dryrun")
## @param $3 test case ID
## @param $4 test case description
## @param $5 Expected exit code.
## @param $6 set 1 if this is not critical (don't care if it's pass or fail)_
function callTestExitEq() {
	callutput=$(. $1 $2)
	retcode=$?
	if (( ${retcode} == $5 ))
	then
		testResult 1 "$3" "$4" $6
	else
		testResult 0 "$3" "$4 ret($retcode)" $6
	fi
}

## @fn callCompareTest()
## @brief Compare two result files expected to be equal
## @param $1 Path to result 1 (golden)
## @param $2 Path to result 2 (test run)
## @param $3 test case ID
## @param $4 test case description
## @param $5 0 if the size is expected to be equal as well. 1 if golden (result 1) might be smaller (will ignore rest of result 2). 2 if the opposite of 1. If $5 > 2, it denotes the max size of compared bytes. (compare the first $5 bytes only)
## @param $6 set 1 if this is not critical (don't care if it's pass or fail)_
function callCompareTest() {
	# Try cmp.
	output=0
	if [[ ! -f "$1" || ! -f "$2" ]]; then
		testResult $output "$3" "$4" $6
		return
	fi
	command -v cmp
	# If cmp is symlink, then it could be from busybox and it does not support "-n" option
	if [[ $? == 0 && ! -L $(which cmp) ]]
	then
		# use cmp
		if (( $5 == 0 )); then
			# Size should be same as well.
			cmp $1 $2
			output=$?
		elif (( $5 == 1 )); then
			# Compare up to the size of golden
			cmp -n `${StatCmd_GetSize} $1` $1 $2
			output=$?
		elif (( $5 == 2 )); then
			# Compare up to the size of test-run
			cmp -n `${StatCmd_GetSize} $2` $1 $2
			output=$?
		else
			# Compare up to $5 bytes.
			cmp -n `${StatCmd_GetSize} $5` $1 $2
			output=$?
		fi
		if (( ${output} == 0 )); then
			output=1
		else
			output=0
		fi
		testResult $output "$3" "$4" $6
	else
	    # use internal logic (slower!)
	    bufsize=`${StatCmd_GetSize} $1`
	    if (( $5 == 2 )); then
		bufsize=`${StatCmd_GetSize} $2`
	    else
		bufsize=$5
	    fi
	    diff <(dd bs=1 count=$bufsize if=$1 &>/dev/null) <(dd bs=1 count=$bufsize if=$2 &>/dev/null)
	    output=$?
	    if (( ${output} == 0 )); then
		output=1
	    else
		output=0
	    fi
	    testResult $output "$3" "$4" $6
	fi
}

## @fn gstTest()
## @brief Execute gst-launch with given arguments
## @todo Separate this function to "gstreamer extension plugin"
## @param $1 gst-launch-1.0 Arguments
## @param $2 test case ID
## @param $3 set 1 if this is not critical (don't care if it's pass or fail)
## @param $4 set 1 if this passes if gstLaunch fails.
## @param $5 set 1 to enable PERFORMANCE test.
## @param $6 set a positive value (seconds) to enable timeout mode.
function gstTest() {
	if [[ "$VALGRIND" -eq "1" ]]; then
		calloutputprefix="valgrind --track-origins=yes ${VALGRIND_SUPPRESSION}"
	fi

	TIMEOUT_AVAIL=1
	if [[ "${6}" -gt "0" ]]; then
		if ! command -v timeout &> /dev/null
		then
			if command -v perl &> /dev/null
			then
				timeout() { perl -e 'alarm shift; exec @ARGV' "$@"; }
			else
				TIMEOUT_AVAIL=0
			fi
		fi
	fi

    if [[ "${6}" -gt "0" && $TIMEOUT_AVAIL -eq 1 ]]; then
		if [[ "${SILENT}" -eq "1" ]]; then
			calloutput=$(eval timeout ${6} $calloutputprefix gst-launch-1.0 -f -q $1 &> /dev/null)
		else
			calloutput=$(eval timeout ${6} $calloutputprefix gst-launch-1.0 -f -q $1)
		fi
	else
		if [[ "${SILENT}" -eq "1" ]]; then
			calloutput=$(eval $calloutputprefix gst-launch-1.0 -f -q $1 &> /dev/null)
		else
			calloutput=$(eval $calloutputprefix gst-launch-1.0 -f -q $1)
		fi
	fi

	retcode=$?
	desired=0
	if [[ "${4}" -eq "1" ]]; then
		if [[ "${retcode}" -ne "0" ]]; then
			desired=1
		fi
	else
		if [[ "${retcode}" -eq "0" ]]; then
			desired=1
		fi
	fi

	if [[ "$desired" -eq "1" ]]; then
		testResult 1 "$2" "gst-launch of case $2" $3
	else
		testResult 0 "$2" "gst-launch of case $2" $3
	fi

	if [[ "$5" -eq "1" ]]; then
		if (( ${#GST_DEBUG_DUMP_DOT_DIR} -le 1 )); then
			GST_DEBUG_DUMP_DOT_DIR="./performance"
		fi
		dot -Tpng $GST_DEBUG_DUMP_DOT_DIR/*.PLAYING_PAUSED.dot > $GST_DEBUG_DUMP_DOT_DIR/debug/$base/$2.png
		gst-report-1.0 --dot $GST_DEBUG_DUMP_DOT_DIR/*.gsttrace | dot -Tsvg > $GST_DEBUG_DUMP_DOT_DIR/profile/$base/$2.svg
		rm -f $GST_DEBUG_DUMP_DOT_DIR/*.dot
		rm -f $GST_DEBUG_DUMP_DOT_DIR/*.gsttrace
	fi
}


## @fn gstTestBackground()
## @brief Execute gst-launch in background with given arguments.
## @param $1 gst-launch-1.0 Arguments, with pipeline description at the end. The pipeline should have async=false for sink elements (skip preroll!).
## @param $2 test case ID
## @param $3 set 1 if this is not critical (don't care if it's pass or fail)
## @param $4 set 1 if this passes if launching the pipeline fails. (pass if timeout expires)
## @param $5 set timeout for launching the pipeline in seconds (not the pipeline EOS). default = 10.
## @return $pid The PID of the background gstreamer pipeline.
function gstTestBackground() {
	local marker=$(mktemp)
	local timeout=10
	local launchSuccess
	local launchFail
	local pipeline

	if [ "$4" == "1" ]; then
		launchSuccess=0
		launchFail=1
	else
		launchSuccess=1
		launchFail=0
	fi
	if [ "$5" == "" ]; then
		# no changes in timeout. do nothing
		sleep 0
	else
		if [ $5 -gt 0 ]; then
			timeout=$5
		else
			# ignore if it's < 1. do nothing
			sleep 0
		fi
	fi

	pipeline="$1   videotestsrc num-buffers=1 ! video/x-raw,width=4,height=4,format=RGB ! filesink location=${marker}"
	gst-launch-1.0 $pipeline &
	pid=$!

	for i in $(seq 1 ${timeout})
	do
		if [ -f "$marker" ]; then
			markersize=$(stat -c%s ${marker})
			if [ $markersize -ge 48 ]; then
				testResult ${launchSuccess} $2 "gst-launch in background of case $2" $3
				rm ${marker}
				return $pid
			fi
		fi
		sleep 1
	done
	rm ${marker}
	testResult ${launchFail} $2 "gst-launch in background of case $2" $3
	return ${pid}
}

## @fn convertBMP2PNG()
## @brief Convert all *.bmp to *.png in the current directory
## @todo macronice "bmp2png" searching.
## @todo Separate this function to "gstreamer extension plugin"
function convertBMP2PNG() {
	tool="bmp2png"
	if [ -x bmp2png ]; then
		tool="bmp2png"
	else
		if [ -x ../bmp2png ]; then
			tool="../bmp2png"
		else
			if [ -x ../../bmp2png ]; then
				tool="../../bmp2png"
			else
				tool="../../../bmp2png"
				# Try this and die if fails
			fi
		fi
	fi
	for X in `ls *.bmp`
	do
		if [[ $X  = *"GRAY8"* ]]; then
			$tool $X --GRAY8
		else
			$tool $X
		fi
	done
}

SSATAPILOADED=1
